File: packages/pyrx_synapse_ios/ios/Classes/PyrxSynapsePlugin.swift
// PyrxSynapsePlugin.swift
// pyrx_synapse_ios — iOS implementation of the PYRX Synapse Flutter SDK.
//
// This file is the FlutterPlugin entry point. Flutter calls
// `register(with:)` at app launch (Flutter's plugin machinery
// auto-generates a `GeneratedPluginRegistrant.m` that invokes us). We do
// three things at registration:
//
//   1. Construct a single `PyrxSynapseHostApi` implementor
//      (`PyrxSynapseHostApiImpl`) and register it via the Pigeon-
//      generated `PyrxSynapseHostApiSetup.setUp`. From this point on,
//      any Dart-side `Synapse.identify(...)` etc. routes through the
//      Pigeon codec to one of the methods on the impl.
//
//   2. Construct a single `StreamEventsStreamHandler` subclass
//      (`PyrxEventStreamHandler`) and register it via the Pigeon-
//      generated `StreamEventsStreamHandler.register`. The handler
//      subscribes to `Pyrx.shared.events()` (the AsyncStream observer
//      surface added in Phase 9.2.1 / PYRXSynapse 0.1.2) the first time
//      Dart attaches, holds the Task as a property, and cancels on
//      detach. One native subscription fans out to all Dart subscribers
//      because Flutter's event channel handles the multi-listener
//      bookkeeping.
//
//   3. Capture cold-start push payload if present (`launchOptions[.remoteNotification]`)
//      and forward it to `Pyrx.shared.recordColdStartLaunch(userInfo:)`.
//      The published SDK's replay buffer of 4 means a Dart subscriber
//      that attaches after the cold-start delivery (the common case
//      because the Flutter engine boots in 0.5-2s while the OS delivers
//      the launch payload synchronously at `application:didFinishLaunching`)
//      still receives the buffered event.
//
// What this file deliberately does NOT do
// ---------------------------------------
// - Swizzle UIApplicationDelegate. The Flutter plugin model gives us
//   first-class hooks (`application(_:didFinishLaunchingWithOptions:)`
//   on FlutterAppDelegate) without monkey-patching. The cost is that
//   the SDK only sees notifications routed through the Flutter
//   delegate; apps that override didReceiveRemoteNotification in their
//   own AppDelegate must call through to super.
// - Implement push permission requests itself. That's delegated to
//   `Pyrx.shared.requestPushPermission(...)` which already handles the
//   UNUserNotificationCenter dance.
// - Handle device-token registration in the plugin layer.
//   `Pyrx.shared.handleDeviceToken(_:)` is called from the customer's
//   AppDelegate `application(_:didRegisterForRemoteNotificationsWithDeviceToken:)`
//   override. PR-2 / customer docs (PR-3) document the wiring; this
//   plugin doesn't try to swizzle around it.

import Flutter
import PYRXSynapse
import UIKit
import UserNotifications

public class PyrxSynapsePlugin: NSObject, FlutterPlugin {

  /// Held strongly for the lifetime of the plugin. The Pigeon setUp call
  /// keeps these alive through the Flutter binary messenger, but holding
  /// them explicitly here makes the lifecycle visible and prevents the
  /// "compiler warning: unused" the Pigeon-setup-without-retain pattern
  /// would otherwise produce.
  private let hostApiImpl: PyrxSynapseHostApiImpl
  private let eventStreamHandler: PyrxEventStreamHandler

  init(messenger: FlutterBinaryMessenger) {
    self.hostApiImpl = PyrxSynapseHostApiImpl()
    self.eventStreamHandler = PyrxEventStreamHandler()
    super.init()
    PyrxSynapseHostApiSetup.setUp(binaryMessenger: messenger, api: hostApiImpl)
    StreamEventsStreamHandler.register(
      with: messenger,
      streamHandler: eventStreamHandler
    )
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let plugin = PyrxSynapsePlugin(messenger: registrar.messenger())
    registrar.publish(plugin)
    registrar.addApplicationDelegate(plugin)
  }

  // ---- UIApplicationDelegate hooks ----
  //
  // These fire on the Flutter delegate (provided the customer's app
  // root inherits from `FlutterAppDelegate`, the default for new Flutter
  // projects). For customers using a custom UIApplicationDelegate, the
  // `addApplicationDelegate` call in `register` makes Flutter forward
  // the lifecycle calls to us anyway.

  public func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil
  ) -> Bool {
    if let userInfo = launchOptions?[.remoteNotification] as? [AnyHashable: Any] {
      Task {
        await Pyrx.shared.recordColdStartLaunch(userInfo: userInfo)
      }
    }
    return true
  }

  public func application(
    _ application: UIApplication,
    didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data
  ) {
    Task {
      do {
        _ = try await Pyrx.shared.handleDeviceToken(deviceToken)
      } catch {
        // The Dart-side caller of `registerForPushNotifications`
        // already received a successful completion; failures here are
        // device-registration races we surface via the SDK's logger
        // (not propagated to Dart because no Dart caller is awaiting
        // this delegate callback).
        NSLog("[PYRXSynapse Flutter] handleDeviceToken failed: \(error)")
      }
    }
  }

  public func application(
    _ application: UIApplication,
    didFailToRegisterForRemoteNotificationsWithError error: Error
  ) {
    Pyrx.shared.handleRegistrationError(error)
  }
}

File: packages/pyrx_synapse_android/android/src/main/kotlin/tech/pyrx/synapse/flutter/PyrxSynapsePlugin.kt
/*
 * PyrxSynapsePlugin.kt
 * pyrx_synapse_android — Android implementation of the PYRX Synapse Flutter SDK.
 *
 * FlutterPlugin entry point. Flutter invokes onAttachedToEngine() at
 * engine-attach time; we do three things there:
 *
 *   1. Construct a single PyrxSynapseHostApi implementor
 *      (PyrxSynapseHostApiImpl) and install it via the Pigeon-generated
 *      PyrxSynapseHostApi.setUp companion. Any Dart-side
 *      `Synapse.identify(...)` etc. routes through the Pigeon codec to
 *      a method on the impl.
 *
 *   2. Construct a single StreamEventsStreamHandler subclass
 *      (PyrxEventStreamHandler) and register it via the Pigeon-generated
 *      StreamEventsStreamHandler.register companion. The handler
 *      lazy-subscribes to Pyrx.events (a SharedFlow from synapse-core
 *      0.1.4's observer surface added in Phase 9.2.1) when the first
 *      Dart listener attaches, and cancels on the last detach. One
 *      coroutine per plugin instance; multiple Dart subscribers are
 *      fanned out by Flutter's EventChannel.
 *
 *   3. Install PyrxPush so the customer's FCM messages route through
 *      the SDK without manifest edits on their side. `PyrxPush.install`
 *      registers the SDK's FCM receiver and wires the push bridge.
 *
 * ActivityAware
 * -------------
 * We implement ActivityAware because requestPushPermission needs to
 * launch ActivityCompat.requestPermissions, which requires an Activity
 * (the OS permission dialog anchors to it). We track the currently
 * attached Activity so the impl can request permission from a foreground
 * surface; if no Activity is attached the impl falls back to reporting
 * "notDetermined" so the caller can retry from a screen.
 */

package tech.pyrx.synapse.flutter

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import tech.pyrx.synapse.flutter.generated.PyrxSynapseHostApi
import tech.pyrx.synapse.flutter.generated.StreamEventsStreamHandler
import tech.pyrx.synapse.push.PyrxPush

class PyrxSynapsePlugin : FlutterPlugin, ActivityAware {

    private var hostApiImpl: PyrxSynapseHostApiImpl? = null
    private var eventStreamHandler: PyrxEventStreamHandler? = null

    // -------------------------------------------------------------------
    // FlutterPlugin
    // -------------------------------------------------------------------

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val appContext = binding.applicationContext

        // Install the synapse-push FCM bridge. Idempotent and safe to
        // call before Pyrx.initialize — the bridge buffers until the
        // SDK is ready (synapse-push 0.1.4 contract).
        try {
            PyrxPush.install(appContext)
        } catch (e: Throwable) {
            // Don't fail plugin attach if the host app's Firebase
            // setup is incomplete; surface a logcat warning and let
            // the customer's debug call discover the issue.
            android.util.Log.w(
                "PYRXSynapseFlutter",
                "PyrxPush.install failed (FCM not configured?): ${e.message}",
            )
        }

        val impl = PyrxSynapseHostApiImpl(appContext)
        val streamHandler = PyrxEventStreamHandler()
        PyrxSynapseHostApi.setUp(binding.binaryMessenger, impl)
        StreamEventsStreamHandler.register(binding.binaryMessenger, streamHandler)

        hostApiImpl = impl
        eventStreamHandler = streamHandler
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        PyrxSynapseHostApi.setUp(binding.binaryMessenger, null)
        eventStreamHandler?.onCancel(null)
        hostApiImpl = null
        eventStreamHandler = null
    }

    // -------------------------------------------------------------------
    // ActivityAware
    // -------------------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        val activity = binding.activity
        hostApiImpl?.setActivity(activity)
        binding.addRequestPermissionsResultListener { requestCode, _, grantResults ->
            hostApiImpl?.onRequestPermissionsResult(requestCode, grantResults) ?: false
        }
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivity() {
        hostApiImpl?.setActivity(null as Activity?)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        hostApiImpl?.setActivity(null as Activity?)
    }
}
